File: impl/java/src/main/java/com/minustwo/chanlun/Canonical.java
// Canonical.java — JSON parser + canonical emitter + SHA-256, matching
// Python's
//
//     json.dumps(obj, sort_keys=True, separators=(comma,colon), ensure_ascii=True)
//
// byte-for-byte over the integer/string domain used by the chanlun-v1 corpus.
//
// Why hand-rolled (no Jackson, no Gson)?
//   - Both libs default to insertion-order output AND tolerate floats; getting
//     them to match Python's `sort_keys + ensure_ascii` exactly is more
//     fragile than writing 200 lines of pure-Java emit. Zero runtime deps
//     keeps the jar tiny and the CI step trivial.
//   - Mirrors impl/go/internal/chanlun/canon.go (the Go port made the same
//     call and it's been green on 48/48 for months).
//
// SHA-equality is the law. A byte off = FAIL.
package com.minustwo.chanlun;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HexFormat;
import java.util.List;

import com.minustwo.chanlun.Types.Kind;
import com.minustwo.chanlun.Types.OrderedObject;
import com.minustwo.chanlun.Types.Value;

public final class Canonical {

    private Canonical() {}

    // ==== Public API =========================================================

    /** Decode JSON bytes into a typed Value tree. Floats are rejected (corpus
     *  is integer-core). */
    public static Value decode(byte[] bytes) {
        Parser p = new Parser(new String(bytes, StandardCharsets.UTF_8));
        Value v = p.parseValue();
        p.skipWhitespace();
        if (!p.atEnd()) {
            throw new IllegalArgumentException(
                "trailing data after JSON value at offset " + p.pos);
        }
        return v;
    }

    /** Canonical JSON bytes:
     *    - object keys sorted lexicographically (code-point order),
     *    - no whitespace,
     *    - non-ASCII chars escaped as backslash-uXXXX (surrogate pair for >0xFFFF). */
    public static byte[] canonical(Value v) {
        StringBuilder sb = new StringBuilder();
        writeValue(sb, v);
        return sb.toString().getBytes(StandardCharsets.UTF_8);
    }

    /** sha256(bytes) as lowercase hex. */
    public static String sha256Hex(byte[] bytes) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(bytes);
            return HexFormat.of().formatHex(digest);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    // ==== Canonical emit =====================================================

    private static void writeValue(StringBuilder sb, Value v) {
        switch (v.kind) {
            case NULL -> sb.append("null");
            case BOOL -> sb.append(v.i != 0 ? "true" : "false");
            case INT -> sb.append(Long.toString(v.i));
            case STRING -> writeString(sb, v.s);
            case ARRAY -> {
                sb.append('[');
                boolean first = true;
                for (Value e : v.arr) {
                    if (!first) sb.append(',');
                    first = false;
                    writeValue(sb, e);
                }
                sb.append(']');
            }
            case OBJECT -> {
                sb.append('{');
                List<String> keys = new ArrayList<>(v.obj.keys());
                Collections.sort(keys); // code-point / UTF-16 order; matches Python for ASCII
                boolean first = true;
                for (String k : keys) {
                    if (!first) sb.append(',');
                    first = false;
                    writeString(sb, k);
                    sb.append(':');
                    writeValue(sb, v.obj.get(k));
                }
                sb.append('}');
            }
        }
    }

    /** Emit a JSON string matching Python json.dumps(..., ensure_ascii=True).
     *  Escapes: quote, backslash, b, f, n, r, t; backslash-u00xx for control bytes < 0x20; raw ASCII
     *  for 0x20..0x7e; backslash-uXXXX for >= 0x7f (surrogate pair for code points
     *  beyond the BMP). */
    private static void writeString(StringBuilder sb, String s) {
        sb.append('"');
        int len = s.length();
        for (int i = 0; i < len; ) {
            int cp = s.codePointAt(i);
            i += Character.charCount(cp);
            switch (cp) {
                case '"' -> sb.append("\\\"");
                case '\\' -> sb.append("\\\\");
                case '\b' -> sb.append("\\b");
                case '\f' -> sb.append("\\f");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                default -> {
                    if (cp < 0x20) {
                        sb.append(String.format("\\u%04x", cp));
                    } else if (cp < 0x7f) {
                        sb.append((char) cp);
                    } else if (cp <= 0xFFFF) {
                        sb.append(String.format("\\u%04x", cp));
                    } else {
                        // Encode as UTF-16 surrogate pair to match Python's
                        // json.dumps(..., ensure_ascii=True) behavior.
                        int u = cp - 0x10000;
                        int high = 0xD800 | ((u >>> 10) & 0x3FF);
                        int low  = 0xDC00 | (u & 0x3FF);
                        sb.append(String.format("\\u%04x\\u%04x", high, low));
                    }
                }
            }
        }
        sb.append('"');
    }

    // ==== Parser =============================================================

    private static final class Parser {
        final String src;
        int pos;

        Parser(String s) { this.src = s; this.pos = 0; }

        boolean atEnd() { return pos >= src.length(); }

        void skipWhitespace() {
            while (pos < src.length()) {
                char c = src.charAt(pos);
                if (c == ' ' || c == '\n' || c == '\r' || c == '\t') {
                    pos++;
                } else {
                    return;
                }
            }
        }

        Value parseValue() {
            skipWhitespace();
            if (pos >= src.length()) throw new IllegalArgumentException("unexpected end of input");
            char c = src.charAt(pos);
            return switch (c) {
                case '{' -> parseObject();
                case '[' -> parseArray();
                case '"' -> Value.ofString(parseString());
                case 't', 'f' -> parseBool();
                case 'n' -> parseNull();
                default -> parseNumber();
            };
        }

        Value parseObject() {
            expect('{');
            OrderedObject o = new OrderedObject();
            skipWhitespace();
            if (peek() == '}') { pos++; return Value.ofObject(o); }
            while (true) {
                skipWhitespace();
                String k = parseString();
                skipWhitespace();
                expect(':');
                Value v = parseValue();
                o.set(k, v);
                skipWhitespace();
                char c = peek();
                if (c == ',') { pos++; continue; }
                if (c == '}') { pos++; return Value.ofObject(o); }
                throw new IllegalArgumentException("expected , or } at " + pos + " got " + c);
            }
        }

        Value parseArray() {
            expect('[');
            List<Value> out = new ArrayList<>();
            skipWhitespace();
            if (peek() == ']') { pos++; return Value.ofArray(out); }
            while (true) {
                out.add(parseValue());
                skipWhitespace();
                char c = peek();
                if (c == ',') { pos++; continue; }
                if (c == ']') { pos++; return Value.ofArray(out); }
                throw new IllegalArgumentException("expected , or ] at " + pos + " got " + c);
            }
        }

        String parseString() {
            expect('"');
            StringBuilder sb = new StringBuilder();
            while (pos < src.length()) {
                char c = src.charAt(pos++);
                if (c == '"') return sb.toString();
                if (c == '\\') {
                    if (pos >= src.length()) throw new IllegalArgumentException("trailing backslash");
                    char esc = src.charAt(pos++);
                    switch (esc) {
                        case '"' -> sb.append('"');
                        case '\\' -> sb.append('\\');
                        case '/' -> sb.append('/');
                        case 'b' -> sb.append('\b');
                        case 'f' -> sb.append('\f');
                        case 'n' -> sb.append('\n');
                        case 'r' -> sb.append('\r');
                        case 't' -> sb.append('\t');
                        case 'u' -> {
                            if (pos + 4 > src.length()) {
                                throw new IllegalArgumentException("short \\u escape");
                            }
                            int hex = Integer.parseInt(src.substring(pos, pos + 4), 16);
                            pos += 4;
                            // Surrogate pair: handle if hex is high surrogate and next is backslash-uXXXX low.
                            if (hex >= 0xD800 && hex <= 0xDBFF
                                && pos + 6 <= src.length()
                                && src.charAt(pos) == '\\'
                                && src.charAt(pos + 1) == 'u') {
                                int low = Integer.parseInt(src.substring(pos + 2, pos + 6), 16);
                                pos += 6;
                                if (low >= 0xDC00 && low <= 0xDFFF) {
                                    int cp = 0x10000 + ((hex - 0xD800) << 10) + (low - 0xDC00);
                                    sb.appendCodePoint(cp);
                                    continue;
                                }
                                // Not a valid low surrogate — fall back to emitting the high as-is.
                                sb.append((char) hex);
                                sb.append((char) low);
                            } else {
                                sb.append((char) hex);
                            }
                        }
                        default -> throw new IllegalArgumentException("bad escape \\" + esc);
                    }
                } else {
                    sb.append(c);
                }
            }
            throw new IllegalArgumentException("unterminated string");
        }

        Value parseBool() {
            if (src.startsWith("true", pos)) { pos += 4; return Value.ofBool(true); }
            if (src.startsWith("false", pos)) { pos += 5; return Value.ofBool(false); }
            throw new IllegalArgumentException("expected bool at " + pos);
        }

        Value parseNull() {
            if (src.startsWith("null", pos)) { pos += 4; return Value.ofNull(); }
            throw new IllegalArgumentException("expected null at " + pos);
        }

        Value parseNumber() {
            int start = pos;
            if (peek() == '-') pos++;
            while (pos < src.length() && Character.isDigit(src.charAt(pos))) pos++;
            // Reject floats: chanlun-v1 corpus is integer-core; a `.`, `e`, or
            // `E` here would mean someone shipped a non-conformant fixture
            // (or a backend leaked a float). Surface as an error.
            if (pos < src.length()) {
                char c = src.charAt(pos);
                if (c == '.' || c == 'e' || c == 'E') {
                    throw new IllegalArgumentException(
                        "chanlun-v1 is integer-core; refusing non-integer number at " + start);
                }
            }
            String num = src.substring(start, pos);
            if (num.isEmpty() || num.equals("-")) {
                throw new IllegalArgumentException("expected number at " + start);
            }
            return Value.ofInt(Long.parseLong(num));
        }

        void expect(char c) {
            if (pos >= src.length() || src.charAt(pos) != c) {
                throw new IllegalArgumentException("expected '" + c + "' at " + pos);
            }
            pos++;
        }

        char peek() {
            if (pos >= src.length()) throw new IllegalArgumentException("unexpected end of input");
            return src.charAt(pos);
        }
    }

    // Convenience: read a file as bytes (used from harness only; here so the
    // tests stay free of java.nio noise).
    public static byte[] readAllBytes(java.nio.file.Path p) {
        try {
            return java.nio.file.Files.readAllBytes(p);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: impl/csharp/src/Chanlun/Canonical.cs
// Canonical JSON encoder + SHA-256, matching Python's
//
//     json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
//
// byte-for-byte over the integer/string domain used by the chanlun-v1 corpus.
//
// System.Text.Json does NOT sort keys lexicographically (it preserves
// insertion order) and its escape policy differs from Python's. We therefore
// roll a custom emitter over an ordered typed Value tree (similar to the Go
// port at impl/go/internal/chanlun/canon.go), then SHA-256 the bytes.

using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Chanlun;

public enum ValueKind { Null, Bool, Int, String, Array, Object }

/// <summary>
/// Typed JSON value tree (deliberately distinguishes int from float — the
/// chanlun-v1 corpus is integer-core).
/// </summary>
public sealed class Value
{
    public ValueKind Kind { get; }
    public long IntVal { get; }
    public bool BoolVal { get; }
    public string StrVal { get; }
    public List<Value> ArrVal { get; }
    public OrderedObject ObjVal { get; }

    private Value(
        ValueKind kind,
        long i = 0,
        bool b = false,
        string? s = null,
        List<Value>? a = null,
        OrderedObject? o = null)
    {
        Kind = kind;
        IntVal = i;
        BoolVal = b;
        StrVal = s ?? string.Empty;
        ArrVal = a ?? new List<Value>();
        ObjVal = o ?? new OrderedObject();
    }

    public static Value Null() => new(ValueKind.Null);
    public static Value Of(bool b) => new(ValueKind.Bool, b: b);
    public static Value Of(long n) => new(ValueKind.Int, i: n);
    public static Value Of(string s) => new(ValueKind.String, s: s);
    public static Value Of(List<Value> a) => new(ValueKind.Array, a: a);
    public static Value Of(OrderedObject o) => new(ValueKind.Object, o: o);

    public List<Value> AsArray()
    {
        if (Kind != ValueKind.Array)
            throw new InvalidOperationException($"AsArray: not an array (kind={Kind})");
        return ArrVal;
    }

    public OrderedObject AsObject()
    {
        if (Kind != ValueKind.Object)
            throw new InvalidOperationException($"AsObject: not an object (kind={Kind})");
        return ObjVal;
    }

    public long AsInt()
    {
        if (Kind != ValueKind.Int)
            throw new InvalidOperationException($"AsInt: not an int (kind={Kind})");
        return IntVal;
    }

    public string AsString()
    {
        if (Kind != ValueKind.String)
            throw new InvalidOperationException($"AsString: not a string (kind={Kind})");
        return StrVal;
    }
}

/// <summary>
/// Insertion-ordered string->Value map. Canonical emit re-sorts keys.
/// </summary>
public sealed class OrderedObject
{
    private readonly List<string> _keys = new();
    private readonly Dictionary<string, Value> _values = new(StringComparer.Ordinal);

    public void Set(string k, Value v)
    {
        if (!_values.ContainsKey(k))
            _keys.Add(k);
        _values[k] = v;
    }

    public bool TryGet(string k, out Value v) => _values.TryGetValue(k, out v!);

    public Value Get(string k) => _values[k];

    public IReadOnlyList<string> Keys => _keys;

    public int Count => _keys.Count;
}

public static class Canonical
{
    /// <summary>
    /// Canonical JSON bytes used for ALL SHA-256 computation. Matches Python
    /// json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
    /// byte-for-byte.
    /// </summary>
    public static byte[] Encode(Value v)
    {
        var sb = new StringBuilder();
        Write(sb, v);
        return Encoding.UTF8.GetBytes(sb.ToString());
    }

    /// <summary>SHA-256 hex digest (lowercase) over `b`.</summary>
    public static string Sha256Hex(byte[] b)
    {
        var sum = SHA256.HashData(b);
        var s = new StringBuilder(sum.Length * 2);
        foreach (var x in sum)
            s.Append(x.ToString("x2", CultureInfo.InvariantCulture));
        return s.ToString();
    }

    private static void Write(StringBuilder sb, Value v)
    {
        switch (v.Kind)
        {
            case ValueKind.Null:
                sb.Append("null");
                break;
            case ValueKind.Bool:
                sb.Append(v.BoolVal ? "true" : "false");
                break;
            case ValueKind.Int:
                sb.Append(v.IntVal.ToString(CultureInfo.InvariantCulture));
                break;
            case ValueKind.String:
                WriteString(sb, v.StrVal);
                break;
            case ValueKind.Array:
                sb.Append('[');
                for (int i = 0; i < v.ArrVal.Count; i++)
                {
                    if (i > 0) sb.Append(',');
                    Write(sb, v.ArrVal[i]);
                }
                sb.Append(']');
                break;
            case ValueKind.Object:
                sb.Append('{');
                // Python json.dumps(sort_keys=True) sorts by raw str key. For ASCII
                // (which the chanlun-v1 corpus uses for keys), Unicode code-point
                // order = lexicographic byte order. StringComparer.Ordinal gives
                // that exact ordering.
                var keys = v.ObjVal.Keys.ToList();
                keys.Sort(StringComparer.Ordinal);
                for (int i = 0; i < keys.Count; i++)
                {
                    if (i > 0) sb.Append(',');
                    WriteString(sb, keys[i]);
                    sb.Append(':');
                    Write(sb, v.ObjVal.Get(keys[i]));
                }
                sb.Append('}');
                break;
            default:
                throw new InvalidOperationException($"canon: unknown kind {v.Kind}");
        }
    }

    /// <summary>
    /// Emit `s` as a JSON string matching Python json.dumps(ensure_ascii=True)
    /// escape rules: backslash, quote, b/f/n/r/t shortcuts; control chars and
    /// non-ASCII via \uXXXX (with UTF-16 surrogate pairs for non-BMP).
    /// </summary>
    private static void WriteString(StringBuilder sb, string s)
    {
        sb.Append('"');
        int i = 0;
        while (i < s.Length)
        {
            char c = s[i];
            switch (c)
            {
                case '"': sb.Append("\\\""); i++; continue;
                case '\\': sb.Append("\\\\"); i++; continue;
                case '\b': sb.Append("\\b"); i++; continue;
                case '\f': sb.Append("\\f"); i++; continue;
                case '\n': sb.Append("\\n"); i++; continue;
                case '\r': sb.Append("\\r"); i++; continue;
                case '\t': sb.Append("\\t"); i++; continue;
            }
            if (c < 0x20)
            {
                sb.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                i++;
                continue;
            }
            if (c < 0x7f)
            {
                sb.Append(c);
                i++;
                continue;
            }
            // ensure_ascii=True path: escape via \uXXXX. C# strings are UTF-16
            // already, so a surrogate pair is two char units — emit both as
            // \uXXXX, exactly like Python json.dumps does.
            if (char.IsHighSurrogate(c) && i + 1 < s.Length && char.IsLowSurrogate(s[i + 1]))
            {
                sb.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                sb.Append("\\u").Append(((int)s[i + 1]).ToString("x4", CultureInfo.InvariantCulture));
                i += 2;
                continue;
            }
            sb.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
            i++;
        }
        sb.Append('"');
    }

    // -- JSON decode --------------------------------------------------------

    /// <summary>
    /// Parse JSON bytes into a typed Value tree. Uses System.Text.Json under
    /// the hood, but reads numbers strictly as long (the chanlun-v1 corpus is
    /// integer-core; a non-integer number anywhere is an error).
    /// </summary>
    public static Value Decode(byte[] bytes)
    {
        var reader = new Utf8JsonReader(
            bytes,
            new JsonReaderOptions
            {
                CommentHandling = JsonCommentHandling.Disallow,
                AllowTrailingCommas = false,
            });
        reader.Read();
        return ReadValue(ref reader);
    }

    private static Value ReadValue(ref Utf8JsonReader reader)
    {
        switch (reader.TokenType)
        {
            case JsonTokenType.Null:
                return Value.Null();
            case JsonTokenType.True:
                return Value.Of(true);
            case JsonTokenType.False:
                return Value.Of(false);
            case JsonTokenType.Number:
                if (!reader.TryGetInt64(out var n))
                    throw new InvalidOperationException(
                        $"Decode: chanlun-v1 is integer-core; non-integer number at byte {reader.TokenStartIndex}");
                return Value.Of(n);
            case JsonTokenType.String:
                return Value.Of(reader.GetString() ?? string.Empty);
            case JsonTokenType.StartArray:
                {
                    var list = new List<Value>();
                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
                        list.Add(ReadValue(ref reader));
                    return Value.Of(list);
                }
            case JsonTokenType.StartObject:
                {
                    var obj = new OrderedObject();
                    while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
                    {
                        if (reader.TokenType != JsonTokenType.PropertyName)
                            throw new InvalidOperationException("Decode: expected property name");
                        var key = reader.GetString() ?? string.Empty;
                        reader.Read();
                        obj.Set(key, ReadValue(ref reader));
                    }
                    return Value.Of(obj);
                }
            default:
                throw new InvalidOperationException($"Decode: unsupported token {reader.TokenType}");
        }
    }
}
